src/components/DateRangeInput/index.tsx

```diff
@@ -0,0 +1,372 @@
+    const [calendarMonthSelectionPopupClassName] = React.useState(randomString(16));
+    const createdContainerRef = React.useRef<HTMLDivElement>(null);
+    const popupRef = React.useRef<HTMLDivElement>(null);
+
+    const containerRef = containerRefFromProps ?? createdContainerRef;
+    const uiModeClassName = useUiModeClassName(uiMode, styles.light, styles.dark);
+    const [
+        showCalendar,
+        setShowCalendarTrue,
+        setShowCalendarFalse,,
+        toggleShowCalendar,
+    ] = useBooleanState(false);
+
+    const hideCalendar = React.useCallback(() => {
+        setTempStartDate(undefined);
+        setShowCalendarFalse();
+    }, [setShowCalendarFalse]);
+
+    const handlePopupBlur = React.useCallback(
+        (isClickedWithin: boolean, e: MouseEvent) => {
+            // Following is to prevent the popup blur when
+            // month selection is changed in the calendar
+            const container = document.getElementsByClassName(
+                calendarMonthSelectionPopupClassName,
+            )[0];
+            const isContainerOrInsideContainer = container
+                ? container === e.target || container.contains(e.target as HTMLElement)
+                : false;
+            if (!isClickedWithin && !isContainerOrInsideContainer) {
+                hideCalendar();
+            }
+        },
+        [hideCalendar, calendarMonthSelectionPopupClassName],
+    );
+
+    useBlurEffect(
+        showCalendar,
+        handlePopupBlur,
+        popupRef,
+        containerRef,
+    );
+
+    const dateRendererParams = React.useCallback(() => ({
+        startDate: tempStartDate ?? value?.startDate,
+        // we only set end date if user hasn't set the start date
+        // i.e. to show previously selected end date)
+        endDate: !tempStartDate ? value?.endDate : undefined,
+    }), [tempStartDate, value]);
+
+    const handleStartDateClick: CalendarProps<never>['onDateClick'] = React.useCallback(
+        (year, month, day) => {
+            setTempStartDate(
+                ymdToDateString(year, month, day),
+            );
+        },
+        [setTempStartDate],
+    );
+
+    const handleEndDateClick: CalendarProps<never>['onDateClick'] = React.useCallback(
+        (year, month, day) => {
+            setTempStartDate((prevTempStartDate) => {
+                if (isDefined(prevTempStartDate) && onChange) {
+                    const lastDate = ymdToDateString(year, month, day);
+                    const prev = new Date(prevTempStartDate).getTime();
+                    const current = new Date(year, month, day).getTime();
+
+                    const startDate = prev > current ? lastDate : prevTempStartDate;
+                    const endDate = prev > current ? prevTempStartDate : lastDate;
+
+                    onChange({
+                        startDate,
+                        endDate,
+                    }, name);
+                }
+
+                return undefined;
+            });
```

Comment:
I still don't like calling sideEffects inside setState. I don't  know if it's the correct thing to do.

Reply:
@AdityaKhatri any thoughts?

Reply:
Maybe try using useReducer and move the side-effects there? Idk if we should call sideEffects there as well

Reply:
Looks like reducer function should also be pure